apps/src/music/views/hooks/useUpdateAnalytics.ts

```diff
@@ -0,0 +1,114 @@
+import {useEffect} from 'react';
+
+import {getCurrentLevel} from '@cdo/apps/code-studio/progressReduxSelectors';
+import {isReadOnlyWorkspace} from '@cdo/apps/lab2/lab2Redux';
+import Lab2Registry from '@cdo/apps/lab2/Lab2Registry';
+import {LifecycleEvent} from '@cdo/apps/lab2/utils/LifecycleNotifier';
+import {useAppSelector} from '@cdo/apps/util/reduxHooks';
+
+import AnalyticsReporter from '../../analytics/AnalyticsReporter';
+
+/**
+ * A hook for updating the {@link AnalyticsReporter} when relevant redux state changes and attaching callbacks
+ * to browser and lifecycle events.
+ */
+function useUpdateAnalytics(analyticsReporter: AnalyticsReporter) {
+  /**
+   * Effect that runs on initial mount
+   *   - Starts a new analytics session.
+   *   - Attaches lifecycle event listeners to start and end sessions when the level changes.
+   *   - Attaches a 'beforeunload' event listener to end the session when the user navigates away from the page.
+   */
+  useEffect(() => {
+    analyticsReporter.startSession();
+
+    const endSession = () => {
+      if (analyticsReporter.isSessionInProgress()) {
+        analyticsReporter.endSession();
+      }
+    };
+
+    // Add lifecycle event listeners to start and end sessions.
+    const lifecycleNotifier = Lab2Registry.getInstance().getLifecycleNotifier();
+    lifecycleNotifier.addListener(
+      LifecycleEvent.LevelLoadCompleted,
+      async (levelProperties, channel) => {
+        if (levelProperties.appName !== 'music') {
+          return;
+        }
+
+        await analyticsReporter.startSession();
+        analyticsReporter.setProjectProperty(
+          'levelType',
+          levelProperties.isProjectLevel ? 'Standalone Project' : 'Level'
+        );
+        analyticsReporter.setProjectProperty('channelId', channel?.id);
+      }
+    );
+
+    lifecycleNotifier.addListener(
+      LifecycleEvent.LevelChangeRequested,
+      endSession
+    );
+
+    // Also attach a listener to end the session if necessary when the level is reloaded without
+    // changing levels, which can happen when the view as user is changed via the Teacher Panel.
+    lifecycleNotifier.addListener(LifecycleEvent.LevelLoadStarted, endSession);
+
+    // TODO: the 'beforeunload' callback is advised against as it is not guaranteed to fire on mobile browsers. However,
+    // we need a way of reporting analytics when the user navigates away from the page. Check with Amplitude for the
+    // correct approach.
+    window.addEventListener('beforeunload', endSession);
```

Comment:
Is this tracked as a follow-up work item?

Reply:
Hmm not sure, I suspect it's been lost in our backlog. I'll create a work item.